arrow-fx-kotlinx-coroutines/src/main/kotlin/arrow/integrations/kotlinx/extensions.kt

```diff
@@ -1,39 +1,105 @@
 package arrow.integrations.kotlinx
 
 import arrow.core.Either
+import arrow.fx.ForIO
+import arrow.fx.IO
 import arrow.fx.IOOf
 import arrow.fx.fix
+import arrow.fx.internal.UnsafePromise
+import arrow.fx.typeclasses.Disposable
+import arrow.fx.typeclasses.Fiber
 import kotlinx.coroutines.CancellationException
 import kotlinx.coroutines.CoroutineScope
 import kotlinx.coroutines.Job
+import kotlinx.coroutines.isActive
 import kotlinx.coroutines.newCoroutineContext
 import kotlinx.coroutines.suspendCancellableCoroutine
 import kotlin.coroutines.EmptyCoroutineContext
 import kotlin.coroutines.resume
 import kotlin.coroutines.resumeWithException
 
-suspend fun <A> IOOf<A>.suspendCancellable(): A = suspendCancellableCoroutine { cont ->
-  val disposable = fix().unsafeRunAsyncCancellable { result ->
-    result.fold<Unit>(cont::resumeWithException, cont::resume)
-  }
+/**
+ * Safely execute the source [IO] in a suspending context,
+ * doing so with [suspendCancellable] makes it participate in structured concurrency.
+ *
+ * This will make sure that the source [IO] is cancelled whenever it's [CoroutineScope] is cancelled.
+ */
+suspend fun <A> IOOf<A>.suspendCancellable(): A =
+  suspendCancellableCoroutine { cont ->
+    if (cont.isActive) {
+      val disposable = fix().unsafeRunAsyncCancellable { result ->
+        result.fold<Unit>(cont::resumeWithException, cont::resume)
+      }
 
-  cont.invokeOnCancellation { disposable() }
-}
+      cont.invokeOnCancellation { disposable() }
+    }
+  }
 
-fun <A> CoroutineScope.unsafeRunIO(io: IOOf<A>, f: (Either<Throwable, A>) -> Unit): Unit =
-  io.unsafeRunScoped(this, f)
+/**
+ * Unsafely run an [IO] and receive the values in a callback [cb] while participating in structured concurrency.
+ * Equivalent of [IO.unsafeRunAsyncCancellable] but with its cancellation token wired to [CoroutineScope].
+ *
+ * @see [IO.unsafeRunAsyncCancellable] for a version that returns the cancellation token instead.
+ */
+fun <A> CoroutineScope.unsafeRunIO(io: IOOf<A>, cb: (Either<Throwable, A>) -> Unit): Unit =
+  io.unsafeRunScoped(this, cb)
 
+/**
+ * Unsafely run an [IO] and receive the values in a callback [cb] while participating in structured concurrency.
+ * Equivalent of [IO.unsafeRunAsyncCancellable] but with its cancellation token wired to [CoroutineScope].
+ *
+ * @see [IO.unsafeRunAsyncCancellable] for a version that returns the cancellation token instead.
+ */
 fun <A> IOOf<A>.unsafeRunScoped(
   scope: CoroutineScope,
-  f: (Either<Throwable, A>) -> Unit
+  cb: (Either<Throwable, A>) -> Unit
```

Comment:
Are these part of the work for homogenizing names across the library?

Reply:
Not really, but that was the reason I renamed here.

So far all callbacks in Arrow Fx have the shape `(Either<Throwable, A>) -> Unit` with param name `cb`.